### DISCUSSION TITLE
Upload vs External Libraries

### DISCUSSION THREAD

**Discussion by jjspierx** [immich-app/immich#5559] [General]
I have been playing around with Immich for about a month now and I am REALLY loving it.  I have taken ~170GB out of Google to import into Immich.  My plan is to be 100% Google free by 2024, and thanks to Immich I can now do this as Photos was my remaining hangup to getting out.

My current Immich setup is using an external library for all of my photos exported from Google, and using the upload library for all new photos going forward.  I guess what I am wondering is if that is a good idea, or if I should upload all of my files into Immich using the upload library.  

For the most part, I don't need to edit/delete the files in the external library, but I would like that option and I don't currently have that with the external library.

The Immich docs mention how external "libraries have a lot of potential for future development beyond what is documented here".  I guess I am curious what some of those developments might be.

Are there other pro's and con's about having my older files be part of an external library with my new files being part of the upload library?  Ideally, I would like a single library, and to be able to edit/delete all of my assets, but it isn't a huge deal not being able to edit/delete my older assets in the external library via Immich.  

Are there any issues with partner sharing that I might not be thinking of related to external libraries?  I plan to use partner sharing with my wife, and want to make sure she is able to view all my assets in both the external and upload libraries.  

Is there any difference in efficiency/resources from Immich perspective whether or not an asset is from upload or external library?

**Comment by jrasm91** [comment]
In my mind it comes down to mostly:

1. How you want to make the files available to the immich containers and
2. How you want to have the files organized on disk.

If you want to use a read-only volume mount you'd have to use external libraries. 
If you already have photos organized on disk by project and want to keep using that structure you'd have to use external libraries.
If you want to upload everything and have immich auto-organize it, use upload library.

The upload libraries have all the features and functionalities of external libraries (and more). External libraries have less features at the expense of being able to use a read-only volume and self-managed file/folder structures. This gap may lessen in the future.

**Reply by jjspierx** [reply]
Thanks @jrasm91 .  That is helpful.  I think I went with external library initially, because it seemed easier to just point to an existing location,, but my volume is not read-only, and I do not need to have a specific organizational structure.  Also, keeping everything in a single library seems easier for backing up, since I don't have to backup multiple locations.  I think I will work on uploading my external library into Immich, and removing the external library.

**Comment by anexas** [comment]
Hi guys. I just started an instance of Immich and successfully connected my photo archive as external library. Everything works perfectly exept for the limitations mentioned above by you. 
I am curious to know if you figured out an easy way to upload in bulk into Immich as against scanning an external library?
Thank you for the information provided by you.

**Reply by nothing2obvi** [reply]
@anexas Immich CLI

**Comment by ai30328348** [comment]
I believe that images should be uploaded to a specific album, not put together in a chaotic manner, and an image can have multiple tags, but not in multiple albums. The image view is too simple, and perhaps a file manager that makes it easy to navigate the images is the best choice.

**Comment by aviv926** [comment]
Starting with [v1.104.0](https://github.com/immich-app/immich/releases/tag/v1.104.0) you can edit/delete files from an external directory if the system administrator has defined this option.

**Comment by teras** [comment]
Following this discussion, I have a question too (new user - already excited about the possibilities!)

I have around 130GB of photos. I managed to import them without problems (I hope). I won't remove photos from the library, my plan is only to add to it.

The problem is that some meta data on the photos are wrong (or missing). And since they are thousands of them, I can only see it per case. Thus, when I spot a wrong meta-data, I want to fix it per case. Immich  as I get it, updates only its own private metadata, doesn't touch the photo itself, right? So, when I need to fix the photo by altering the EXIF information, that won't be possible using uploaded images.

Using external libraries as I get it this is possible. The only difference might be that the checksum of the file will change, but I don't find it as a problem since the new metadata will be the "correct" one. Is this the only solution or there's a way to update file metadata through a plugin, maybe?  

Also another thing is about new photos. These photos will be uploaded through the mobile app interface. So they'll be put on another place on disk, right? What if I want to merge these locations, and have only one location, e.g. move the uploaded files later on to the external library location? is it possible?

**Reply by ilopmar** [reply]
I have been using Immich for a year now (since v1.105) and I only use external libraries. I have two different types of external libraries:
- One with a folder per day and event.
- Another one for every one of us at home for the pictures we take using our phones.

For the former (for example when we go to a trip), I review and clean all the duplicated, wrong,... photos and then store in a folder with the date. That volume is mounted as read-only in Immich.
Then, for our phone pictures, I synchronize them to my NAS to a "user/year" folder structure, and then every night I run a script that refresh Immich external libraries if there is a change. Those volumes are mounted as read-write because I want that everyone at home be able to delete/edit it's own photos.

This way all the pictures and videos are stored in the directory structure I want (useful for backups) and if something goes wrong I only need to reprocess them but I don't lost anything just because I uploaded it directly from my phone to Immich and the photo "gets lost" in Immich internal directory structure.

**Reply by teras** [reply]
> I have been using Immich for a year now (since v1.105) and I only use external libraries. I have two different types of external libraries:

It looks like the only serious solution is to have the photos in external libraries. Thank you for the hints. Only one question:

> ... if something goes wrong I only need to reprocess them

what about the meta data, and especially the face recognition? If I start from scratch, all the effort to clean up the faces (remove hundreds of irrelevant faces, merges etc) is lost. Or should I keep the meta data as is, replace the files and let immich do its magic?

**Reply by ilopmar** [reply]
> what about the meta data, and especially the face recognition? If I start from scratch, all the effort to clean up the faces (remove hundreds of irrelevant faces, merges etc) is lost. Or should I keep the meta data as is, replace the files and let immich do its magic?

I haven't done it but it something is really wrong I'd need to reprocess everything and lose all faces, metadata and things like that. Probably I'd try first just reprocessing and cross my fingers because maybe Immich fixes everything.
In any case, in all those upgrades I've done in the last year everything has worked properly (even with some breaking changes), so I think I'll be good.

**Reply by teras** [reply]
Thank you
